client/src/core/model-registry.js

```diff
@@ -218,6 +223,78 @@ class ModelRegistry {
     this.tmpMaterials = [];
   }
 
+  /**
+   * Applies multiple volume-preserving skews to geometry.
+   * @param {BufferGeometry} geometry - The geometry to skew.
+   * @param {number} [p1] - Shear factor for Z from Y.
+   * @param {number} [p2] - Shear factor for X from Z.
+   * @param {number} [p3] - Shear factor for Y from X.
+   * @param {number} [p4] - Shear factor for Y from Z.
+   * @param {number} [p5] - Shear factor for Z from X.
+   * @param {number} [p6] - Shear factor for X from Y.
+   */
+  applyShear(geometry, p1 = 0, p2 = 0, p3 = 0, p4 = 0, p5 = 0, p6 = 0) {
+    const shears = [
+      {target: 'z', source: 'y', factor: -clampShear(p1)},
+      {target: 'x', source: 'z', factor: clampShear(p2)},
+      {target: 'y', source: 'x', factor: -clampShear(p3)},
+      {target: 'y', source: 'z', factor: clampShear(p4)},
+      {target: 'z', source: 'x', factor: -clampShear(p5)},
+      {target: 'x', source: 'y', factor: clampShear(p6)},
+    ];
+
+    const axes = {x: 0, y: 1, z: 2};
+
+    for (const {target, source} of shears) {
+      if (!axes.hasOwnProperty(target) ||
+        !axes.hasOwnProperty(source) ||
+        target === source) {
+        console.warn('Invalid shear axis pair:', target, source);
+        return geometry;
+      }
+    }
+
+    geometry = geometry.clone();
+    geometry.computeBoundingBox();
+    const center = new Vector3();
+    geometry.boundingBox.getCenter(center);
+
+    const posAttr = geometry.attributes.position;
+    const vertexOriginal = new Vector3();
+    const vertexModified = new Vector3();
+
+    const originalPositions = [];
+    for (let i = 0; i < posAttr.count; i++) {
+      vertexOriginal.fromBufferAttribute(posAttr, i);
+      originalPositions.push(vertexOriginal.clone());
+    }
+
+    for (let i = 0; i < posAttr.count; i++) {
+      const original = originalPositions[i];
+      vertexModified.copy(original);
+
+      for (const {target, source, factor} of shears) {
+        const tIdx = axes[target];
+        const sIdx = axes[source];
+        const sourceValue = original
+            .getComponent(sIdx) - center.getComponent(sIdx);
+        const offset = sourceValue * factor;
+        vertexModified.setComponent(
+            tIdx,
+            vertexModified.getComponent(tIdx) + offset,
+        );
+      }
```

Comment:
Couldn't this be only one loop?

What the first loop does is store the original position in an array, then the second ones fetches them one by one, but then might as well be one for loop like this:

```javascript
const original = new Vector3(); // rename `vertexOriginal` into `original`

...

for (let i = 0; i < posAttr.count; i++) {
    original.fromBufferAttribute(posAttr, i);
    vertexModified.copy(original);
    ...
}
```